<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Manipulate Taxonomic Composition Data of Microbial Communities • tidytacos</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Manipulate Taxonomic Composition Data of Microbial Communities">
<meta name="description" content="A toolkit for processing and visualizing taxonomic composition data of microbial communities, based on the tidyverse framework.">
<meta property="og:description" content="A toolkit for processing and visualizing taxonomic composition data of microbial communities, based on the tidyverse framework.">
<meta property="og:image" content="https://lebeerlab.github.io/tidytacos/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">tidytacos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="articles/index.html">Articles</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/LebeerLab/tidytacos/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="tidytacos-">tidytacos <a class="anchor" aria-label="anchor" href="#tidytacos-"></a>
</h1>
</div>
<p><a href="https://github.com/LebeerLab/tidytacos/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/LebeerLab/tidytacos/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a> <a href="https://codecov.io/gh/LebeerLab/tidytacos" class="external-link"><img src="https://codecov.io/gh/LebeerLab/tidytacos/graph/badge.svg?token=532YS16DXU" alt="codecov"></a> <br><br><br><br><br></p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Tidytacos (tidy TAxonomic COmpositionS) is an R package for the exploration of microbial community data. Such community data consists of read counts generated by amplicon sequencing (e.g. a region of the 16S rRNA gene) or metagenome (shotgun) sequencing. Each read count represents a number of sequencing reads identified for some taxon (an ASV, OTU, species, or higher-level taxon) in a sample.</p>
<p>Tidytacos builds on the <a href="https://www.tidyverse.org/" class="external-link">tidyverse</a> created by <a href="http://hadley.nz/" class="external-link">Hadley Wickham</a>: the data are stored in tidy tables where each row is an observation and each column a variable. In addition, the package supplies a set of “verbs”: functions that take a tidytacos object as first argument and also return a tidytacos object. This makes it easy to construct “pipe chains” of code that represent series of operations performed on the tidytacos object.</p>
</div>
<div class="section level2">
<h2 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<p>Tidytacos is an R package. You can find instructions to download and install R <a href="https://cran.r-project.org/" class="external-link">here</a>.</p>
<p>Tidytacos relies on the tidyverse R package (or, more accurately, set of R packages). You can install the tidyverse by running the following R code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code></pre></div>
<p>Finally, RStudio is a nice IDE to work with R code (as well as code in other scripting languages). It has a lot more features than what the default R IDE allows: beyond creating and saving scripts, it also shows your figures, allows you to navigate files, allows you to inspect tables etc. You can download RStudio <a href="https://posit.co/downloads/" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Run the following R code to install the latest version of tidytacos:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"LebeerLab/tidytacos"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>If your ASVs are counted and annotated using <a href="https://benjjneb.github.io/dada2/" class="external-link">dada2</a>, you can use the following function to convert the results to a tidytacos object:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taco</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/from_dada.html">from_dada</a></span><span class="op">(</span><span class="va">seqtab.nochim</span>, <span class="va">taxa</span><span class="op">)</span></span></code></pre></div>
<p>If you have data in the form of a phyloseq object you could convert it using:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taco</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/from_phyloseq.html">from_phyloseq</a></span><span class="op">(</span><span class="va">phylo_obj</span><span class="op">)</span></span></code></pre></div>
<p>Where seqtab.nochim and taxa refer to the R objects <a href="https://benjjneb.github.io/dada2/tutorial.html" class="external-link">as calculated in the dada2 tutorial</a>.</p>
<p>You may wish to create a tidytacos object from your counts matrix, for example an OTU table where rownames are taxa and colnames are samples. After that is done, you can add your taxonomy table and sample data. The variable ‘taxon’ of the taxonomy table should align with the rownames of the OTU table. Furthermore it may include all taxonomic levels from ‘kingdom’ to ‘species’ and a ‘sequence’ variable (nucleotide sequence). The variable ‘sample’ of the sample data should align with the colnames of the OTU table.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taco</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/create_tidytacos.html">create_tidytacos</a></span><span class="op">(</span><span class="va">OTUtable</span>, taxa_are_columns <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">taco</span> <span class="op">&lt;-</span> <span class="va">taco</span><span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/add_metadata.html">add_metadata</a></span><span class="op">(</span><span class="va">taxonomy</span>, table_type<span class="op">=</span><span class="st">"taxa"</span><span class="op">)</span><span class="op"><a href="reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="reference/add_metadata.html">add_metadata</a></span><span class="op">(</span><span class="va">sampledata</span>, table_type<span class="op">=</span><span class="st">"sample"</span><span class="op">)</span></span></code></pre></div>
<p>A tidytacos object is read and stored as three sparse tables (counts-, taxa- and samples.csv). To read in existing data from a folder, <a href="https://github.com/LebeerLab/tidytacos/tree/dev/data-raw/tidytacos/leaf" class="external-link">for example one called ‘leaf’ in the ‘data-raw/tidytacos’ folder</a> you would run:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taco</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/read_tidytacos.html">read_tidytacos</a></span><span class="op">(</span><span class="st">"data-raw/tidytacos/leaf"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h2>
<p><a href="https://lebeerlab.github.io/tidytacos/reference/index.html">A documentation page (help page)</a> is available for all functions in the browser or in R. You can view it in R by running e.g. <code><a href="reference/filter_samples.html">?filter_samples</a></code>. Some useful tutorials can be found on the <a href="https://github.com/LebeerLab/tidytacos/wiki" class="external-link">wiki</a>.</p>
</div>
<div class="section level2">
<h2 id="need-support">Need support?<a class="anchor" aria-label="anchor" href="#need-support"></a>
</h2>
<p>Post on <a href="https://github.com/LebeerLab/tidytacos/issues" class="external-link">GitHub issues</a> if you have questions, requests, or if you run into an issue.</p>
</div>
<div class="section level2">
<h2 id="feel-like-contributing">Feel like contributing?<a class="anchor" aria-label="anchor" href="#feel-like-contributing"></a>
</h2>
<p>Please read the <a href="https://github.com/LebeerLab/tidytacos/wiki/Developer-Guide" class="external-link">GitHub Developer Guide</a>. Fork the dev branch, make your changes and make a pull request. Your suggestions will be reviewed and if approved, will be implemented in the next release.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/LebeerLab/tidytacos/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/LebeerLab/tidytacos/issues/new?assignees=TheOafidian&amp;labels=bug&amp;projects=&amp;template=bug_report.md&amp;title=%5BBUG%5D+" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing tidytacos</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Stijn Wittouck <br><small class="roles"> Author, maintainer </small>  </li>
<li>Tim Van Rillaer <br><small class="roles"> Author </small>  </li>
<li>Wenke Smets <br><small class="roles"> Author </small>  </li>
<li>Sarah Lebeer <br><small class="roles"> Funder </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Stijn Wittouck, Tim Van Rillaer, Wenke Smets, Sarah Lebeer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
