---
title: "Common workflows with tidyamplicons"
author: "Stijn Wittouck"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Common workflows}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F, fig.width = 8, fig.height = 6)
```

This vignette explains how to perform some common amplicon count data analysis tasks using tidyamplicons. I will illustrate those using a dataset with human microbiome samples from the upper respiratory tract (URT), taken from [this paper](https://www.frontiersin.org/articles/10.3389/fmicb.2017.02372/full) by De Boeck et al. It contains nose as well as nasopharynx samples. Most samples were taken using a swab method, but a minority was taking with the aspirate method. 

# Philosophy of the package

A tidyamplicons object is simply a list of three tables:

* **abundances**: These are the counts of reads for each taxon (OTU/ASV/phylotype) in each sample. Each row represents such a read count. 
* **samples**: This table contains the sample metadata. Each row of course represents a sample. 
* **taxa**: This table contains the taxonomy and other metadata for the taxa. Each row represents a taxon. 

The package is called **tidy**amplicons because each of the tables is tidy: each row represents an observation and each column a variable. 

The main differences with the popular phyloseq package are:

* The otu_table object of phyloseq is not tidy, while the abundances table of tidyamplicons is. This tidiness sacrifices some memory and speed, while in return gaining a data format that is easier to process and visualize with only a few lines of code. 
* All tables in tidyamplicons are just that: tables (or "tibbles", so you prefer). In phyloseq, each table has its own special data type (i.e. the classes "otu_table", "tax_table" and "sample_data"). This makes it sometimes difficult to access the data directly and do simple stuff with it. Also, the otu_table and tax_table classes are based on the matrix type, while the sample_data class is based on the dataframe type. This also makes data manipulation sometimes unstraightforward. 
* In a phyloseq otu_table, sometimes the rows are the taxa and sometimes the columns. This can easily leads to errors. 

# Common workflows

We need only three packages: tidyamplicons (of course) and the tidyverse packages dplyr (table manipulation) and ggplot2 (visualization). 

```{r}
library(tidyamplicons)
library(dplyr)
library(ggplot2)
```

Let's load our example data set and inspect the samples table: 

```{r}
load("../data/urt.rda")
glimpse(urt$samples)
```

## Making a barplot of a subset of samples

We can very easily create an explorative plot of a subset of our samples (e.g., only nose samples taken with the swab method) in the following way: 

```{r}
urt %>%
  filter_samples(location == "N", method == "S") %>%
  get_bar_plot()
```

The `filter_samples` function does what it says: filtering samples. It will also delete taxa from the taxa table that have zero total reads in the remaining samples. The `get_bar_plot` function returns a nice barplot visualization of the most abundant taxa in our samples. In addition, it does some things behind the screens: 

* Add relative abundances to the abundances table. 
* For each taxon, calculate the maximum relative abundance across all samples where this taxon occurs. 
* Give all taxa a human-understandable name so that the taxon name is unique. This is currently just the genus name of the taxon, followed by a number to make it unique. Taxa with a higher maximum relative abundance get a smaller number. E.g. "Lactobacillus 1" is the Lactobacillus taxon with the largests maximum relative abundance. 
* Make a new variable that is equal to the taxon name, except that only the top-11 taxa retain there name and all others are changed to "residual". This for visualization purposes; the human eye can only discriminate clearly between about 12 colors. 

All these new variables are created under the hood, but are gone when the function execution is finished (there are no so-called "side-effects"). Luckily, for each of these variables there also exists a function that will create it and keep it! These functions are the following (there names are intended to be self-explanatory as much as possible): 

* `add_rel_abundance`: adds to abundances table
* `add_max_rel_abundance`: adds to taxa table
* `add_taxon_name`: adds to taxa table
* `add_taxon_name_color`: adds to taxa table

There are more of such functions that add an extra variable to one of the tables; we will explore them further in other parts of this vignette. 

## Making a facet wrapped barplot

We start by selecting the samples we are interesting in: only the ones taken with the swab method. 

```{r}
urt_subset <- filter_samples(urt, method == "S")
```

There are two interesting types of facet wrapped barplots we can make: facets with fixed x-axis categories, and facets witout fixed x-axis categories. 

### Facets with fixed x-axis categories

The first type is a plot where there is some variable on the x-axis that is not the sample name (the not-very-informative default). We want the categories of that variable in the same order in all our subplots. For example, we have a subplot for every participant, and for each participant we want to see the nose sample and the nasopharynx sample, in that order. This is achieved by adding a `x = location` extra argument to the barplot function, and adding a `facet_wrap()` layer. Putting the x-axis categories in the same order in all subplots is the default behaviour of the `facet_wrap()` function. Note that the order of the samples in each subplot is now determined by the categories of the variable we put there (e.g. alphabetically sorted), and no longer by a sample clustering procedure! 

```{r, fig.height = 12}
get_bar_plot(urt_subset, x = location) +
  facet_wrap(~ participant, nrow = 10)
```

### Facets without fixed x-axis categories

The second type of facet wrapped plot is one where we have the sample names on the x-axis, as usual, and they are sorted according to a clustering procedure. For example, we want one facet per sampling location, with only the samples belonging to that location. This can be achieved in the following way:

```{r}
get_bar_plot(urt_subset) +
  facet_wrap(~ location, nrow = 2)
```

This is not quite right. The default behavior of `facet_wrap()` is to repeat all possible x-axis values in all facets, even if there is no information there! In our case, the default behavior would be to put all samples names in the nose facet and nasopharynx facet, and plot empty space if a sample name - sampling location combination doesn't exist. Adding the argument `scales = "free_x"` corrects this behavior and also makes sure that the samples are plotted in order of clustering:

```{r}
get_bar_plot(urt_subset) +
  facet_wrap(~ location, scales = "free_x", nrow = 2)
```

## Plotting only a subset of taxa, without "residual"

To be able to do this, we need to do a number of things by hand that normally happen "under the hood" in the `get_bar_plot()` function:

* Step 1: We add relative abundances. We want them to be calculated with repect to the full samples, before the taxa we're not interested in are removed! 
* Step 2: We select only the taxa we want (e.g. family Lactobacillaceae). 
* Step 3: We add the variable "taxon_name_color"; this variable is equal to "taxon_name", except that everything apart from the n most abundant taxa (which are at this moment all Lactobacillaceae!) will be called "residual". 
* Step 4: We make the plot. 

Thanks to the `%>%` (pipe) operator from the magrittr package, we can achieve al this using the following elegant code: 

```{r}
urt %>%
  add_rel_abundance() %>%
  filter_taxa(family == "Lactobacillaceae") %>%
  add_taxon_name_color() %>%
  get_bar_plot()
```
