<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ tidytacos</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://lebeerlab.github.io/tidytacos/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidytacos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/LebeerLab/tidytacos/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="tidytacos">Tidytacos<a class="anchor" aria-label="anchor" href="#tidytacos"></a></h2>

      <div class="section-desc"><p>Package overview</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="tidytacos-package.html">tidytacos-package</a></code>

        </dt>
        <dd>tidytacos: Manipulate Taxonomic Composition Data of Microbial Communities</dd>
      </dl><dl><dt>

          <code><a href="tidytacos.html">tidytacos</a></code>

        </dt>
        <dd>tidytacos: Functions to manipulate and visualize amplicon count data.</dd>
      </dl></div><div class="section level2">
      <h2 id="file-handling">File handling<a class="anchor" aria-label="anchor" href="#file-handling"></a></h2>

      <div class="section-desc"><p>Functions for loading in or creating tidytaco objects or converting to another datatype.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="read_tidytacos.html">read_tidytacos()</a></code>

        </dt>
        <dd>Read community data written by tidytacos</dd>
      </dl><dl><dt>

          <code><a href="write_tidytacos.html">write_tidytacos()</a></code>

        </dt>
        <dd>Write community data in tidytacos format</dd>
      </dl><dl><dt>

          <code><a href="merge_tidytacos.html">merge_tidytacos()</a></code>

        </dt>
        <dd>Merge two tidytacos objects</dd>
      </dl><dl><dt>

          <code><a href="create_tidytacos.html">create_tidytacos()</a></code>

        </dt>
        <dd>Initiate tidytacos object</dd>
      </dl><dl><dt>

          <code><a href="add_metadata.html">add_metadata()</a></code>

        </dt>
        <dd>Add metadata to the tidytacos object</dd>
      </dl><dl><dt>

          <code><a href="from_dada.html">from_dada()</a></code>

        </dt>
        <dd>DADA2 to a tidytacos object</dd>
      </dl><dl><dt>

          <code><a href="from_phyloseq.html">from_phyloseq()</a></code>

        </dt>
        <dd>Convert phyloseq object to tidytacos object</dd>
      </dl><dl><dt>

          <code><a href="as_phyloseq.html">as_phyloseq()</a></code>

        </dt>
        <dd>Convert tidytacos object to phyloseq object</dd>
      </dl><dl><dt>

          <code><a href="to_biom.html">to_biom()</a></code>

        </dt>
        <dd>Write the counts of the tidytacos object to a biom file</dd>
      </dl><dl><dt>

          <code><a href="to_fasta.html">to_fasta()</a></code>

        </dt>
        <dd>Write the sequences of the taxa table to a fasta file</dd>
      </dl></div><div class="section level2">
      <h2 id="table-manipulation">Table manipulation<a class="anchor" aria-label="anchor" href="#table-manipulation"></a></h2>

      <div class="section-desc"><p>Functions for manipulation of the three distinct tables using tidy-related functions.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="samples.html">samples()</a></code>

        </dt>
        <dd>Extract the sample table</dd>
      </dl><dl><dt>

          <code><a href="taxa.html">taxa()</a></code>

        </dt>
        <dd>Extract the taxon table</dd>
      </dl><dl><dt>

          <code><a href="counts.html">counts()</a></code>

        </dt>
        <dd>Extract the count table</dd>
      </dl><dl><dt>

          <code><a href="select_counts.html">select_counts()</a></code>

        </dt>
        <dd>Retain or remove a set of count variables</dd>
      </dl><dl><dt>

          <code><a href="select_samples.html">select_samples()</a></code>

        </dt>
        <dd>Retain or remove a set of sample variables</dd>
      </dl><dl><dt>

          <code><a href="select_taxa.html">select_taxa()</a></code>

        </dt>
        <dd>Retain or remove a set of taxon variables</dd>
      </dl><dl><dt>

          <code><a href="mutate_counts.html">mutate_counts()</a></code>

        </dt>
        <dd>Create extra variables in the count table</dd>
      </dl><dl><dt>

          <code><a href="mutate_samples.html">mutate_samples()</a></code>

        </dt>
        <dd>Create extra variables in the sample table</dd>
      </dl><dl><dt>

          <code><a href="mutate_taxa.html">mutate_taxa()</a></code>

        </dt>
        <dd>Create extra variables in the taxa table</dd>
      </dl><dl><dt>

          <code><a href="filter_counts.html">filter_counts()</a></code>

        </dt>
        <dd>Filter the counts</dd>
      </dl><dl><dt>

          <code><a href="filter_samples.html">filter_samples()</a></code>

        </dt>
        <dd>Filter the samples</dd>
      </dl><dl><dt>

          <code><a href="filter_taxa.html">filter_taxa()</a></code>

        </dt>
        <dd>Filter the taxa</dd>
      </dl><dl><dt>

          <code><a href="aggregate_samples.html">aggregate_samples()</a></code>

        </dt>
        <dd>Aggregate samples with identical values for all metadata</dd>
      </dl><dl><dt>

          <code><a href="aggregate_taxa.html">aggregate_taxa()</a></code>

        </dt>
        <dd>Aggregate taxa on a given taxonomic rank</dd>
      </dl><dl><dt>

          <code><a href="everything.html">everything()</a></code>

        </dt>
        <dd>Get all data in one single table</dd>
      </dl></div><div class="section level2">
      <h2 id="relative-abundance-prevalence-and-other-count-computations">Relative abundance, prevalence and other count computations<a class="anchor" aria-label="anchor" href="#relative-abundance-prevalence-and-other-count-computations"></a></h2>

      <div class="section-desc"><p>Functions for transforming the count table to different matrix representations or pairwise comparison.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="counts_matrix.html">counts_matrix()</a></code>

        </dt>
        <dd>Return a counts matrix</dd>
      </dl><dl><dt>

          <code><a href="rel_abundance_matrix.html">rel_abundance_matrix()</a></code>

        </dt>
        <dd>Return a relative abundance matrix</dd>
      </dl><dl><dt>

          <code><a href="add_rel_abundance.html">add_rel_abundance()</a></code>

        </dt>
        <dd>Add relative abundances to count table</dd>
      </dl><dl><dt>

          <code><a href="add_mean_rel_abundance.html">add_mean_rel_abundance()</a></code>

        </dt>
        <dd>Add average relative abundances to taxa table</dd>
      </dl><dl><dt>

          <code><a href="mean_rel_abundances.html">mean_rel_abundances()</a></code>

        </dt>
        <dd>Get mean relative abundances of taxa in general or per condition</dd>
      </dl><dl><dt>

          <code><a href="add_absolute_abundance.html">add_absolute_abundance()</a></code>

        </dt>
        <dd>Add absolute abundances to count table</dd>
      </dl><dl><dt>

          <code><a href="add_spike_ratio.html">add_spike_ratio()</a></code>

        </dt>
        <dd>Add spike ratio</dd>
      </dl><dl><dt>

          <code><a href="add_density.html">add_density()</a></code>

        </dt>
        <dd>Add density to count table</dd>
      </dl><dl><dt>

          <code><a href="add_logratio.html">add_logratio()</a></code>

        </dt>
        <dd>Add logratios</dd>
      </dl><dl><dt>

          <code><a href="add_clr_abundance.html">add_clr_abundance()</a></code>

        </dt>
        <dd>Perform a centered log ratio transformation on the readcounts.</dd>
      </dl><dl><dt>

          <code><a href="add_jervis_bardy.html">add_jervis_bardy()</a></code>

        </dt>
        <dd>Apply the taxon QC method of Jervis-Bardy</dd>
      </dl><dl><dt>

          <code><a href="prevalences.html">prevalences()</a></code>

        </dt>
        <dd>Get prevalences of taxa in general or per condition</dd>
      </dl><dl><dt>

          <code><a href="add_prevalence.html">add_prevalence()</a></code>

        </dt>
        <dd>Add taxon prevalences to the taxon table</dd>
      </dl><dl><dt>

          <code><a href="counts_tidy.html">counts_tidy()</a></code>

        </dt>
        <dd>Convert matrix with counts to tidy data frame</dd>
      </dl><dl><dt>

          <code><a href="add_total_absolute_abundance.html">add_total_absolute_abundance()</a></code>

        </dt>
        <dd>Add total absolute abundances of samples</dd>
      </dl><dl><dt>

          <code><a href="add_total_count.html">add_total_count()</a></code>

        </dt>
        <dd>Add total read count per sample</dd>
      </dl><dl><dt>

          <code><a href="add_total_density.html">add_total_density()</a></code>

        </dt>
        <dd>Add total densities of samples</dd>
      </dl></div><div class="section level2">
      <h2 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a></h2>

      <div class="section-desc"><p>Functions for various plots that can be made with the data in a tidytacos object.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="tacoplot_alphas.html">tacoplot_alphas()</a></code>

        </dt>
        <dd>Return a boxplot of every alpha metric per group in the samples table of a tidytaco object. If no alpha metrics are present, all available ones are added.</dd>
      </dl><dl><dt>

          <code><a href="tacoplot_codifab.html">tacoplot_codifab()</a></code>

        </dt>
        <dd>Generate a compositional differential abundance plot</dd>
      </dl><dl><dt>

          <code><a href="tacoplot_euler.html">tacoplot_euler()</a></code>

        </dt>
        <dd>Return an euler diagram of overlapping taxon_ids between conditions</dd>
      </dl><dl><dt>

          <code><a href="tacoplot_ord.html">tacoplot_ord()</a></code>

        </dt>
        <dd>Return an ordination plot of the samples</dd>
      </dl><dl><dt>

          <code><a href="tacoplot_ord_ly.html">tacoplot_ord_ly()</a></code>

        </dt>
        <dd>Return an interactive ordination plot of the samples</dd>
      </dl><dl><dt>

          <code><a href="tacoplot_prevalences.html">tacoplot_prevalences()</a></code>

        </dt>
        <dd>Return a heatmap of all taxa above a certain threshold prevalence per condition, clusters them and compares prevalences with a fisher_test.</dd>
      </dl><dl><dt>

          <code><a href="tacoplot_scree.html">tacoplot_scree()</a></code>

        </dt>
        <dd>Return a scree plot to visualize the eigenvalues of the PCA.</dd>
      </dl><dl><dt>

          <code><a href="tacoplot_stack.html">tacoplot_stack()</a></code>

        </dt>
        <dd>Return a bar plot of the samples</dd>
      </dl><dl><dt>

          <code><a href="tacoplot_stack_ly.html">tacoplot_stack_ly()</a></code>

        </dt>
        <dd>Return an interactive bar plot of the samples</dd>
      </dl><dl><dt>

          <code><a href="tacoplot_venn.html">tacoplot_venn()</a></code>

        </dt>
        <dd>Return a venn diagram of overlapping taxon_ids between conditions</dd>
      </dl><dl><dt>

          <code><a href="tacoplot_venn_ly.html">tacoplot_venn_ly()</a></code>

        </dt>
        <dd>Return an interactive venn diagram of overlapping taxon_ids between conditions</dd>
      </dl><dl><dt>

          <code><a href="tacoplot_zoom.html">tacoplot_zoom()</a></code>

        </dt>
        <dd>Return a visualization designed for a small number of samples</dd>
      </dl></div><div class="section level2">
      <h2 id="statistical-tests">Statistical tests<a class="anchor" aria-label="anchor" href="#statistical-tests"></a></h2>

      <div class="section-desc"><p>Functions for various statistical tests that can be performed on a tidytacos object.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="perform_adonis.html">perform_adonis()</a></code>

        </dt>
        <dd>Perform an adonis test</dd>
      </dl><dl><dt>

          <code><a href="perform_anosim.html">perform_anosim()</a></code>

        </dt>
        <dd>Perform anosim test</dd>
      </dl><dl><dt>

          <code><a href="perform_mantel_test.html">perform_mantel_test()</a></code>

        </dt>
        <dd>Determine the correlation between the distance of the counts in a tidytacos object and a sample variable, multiple sample variables or another matrix.</dd>
      </dl><dl><dt>

          <code><a href="add_codifab.html">add_codifab()</a></code>

        </dt>
        <dd>Perform compositional differential abundance analysis</dd>
      </dl></div><div class="section level2">
      <h2 id="taxonomy-related-functions">Taxonomy related functions<a class="anchor" aria-label="anchor" href="#taxonomy-related-functions"></a></h2>

      <div class="section-desc"><p>Functions to manipulate taxonomy names.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="rank_names.html">rank_names()</a></code>

        </dt>
        <dd>Return rank names associated with a tidytacos object if these are defined. In case of undefined rank names, the function returns the taxon_id field.</dd>
      </dl><dl><dt>

          <code><a href="set_rank_names.html">set_rank_names()</a></code>

        </dt>
        <dd>Set rank names for a tidytacos object</dd>
      </dl><dl><dt>

          <code><a href="add_taxon_name.html">add_taxon_name()</a></code>

        </dt>
        <dd>Add sensible taxon name to taxon table</dd>
      </dl><dl><dt>

          <code><a href="add_taxon_name_color.html">add_taxon_name_color()</a></code>

        </dt>
        <dd>Add taxon color for visualization.</dd>
      </dl><dl><dt>

          <code><a href="add_eigentaxa.html">add_eigentaxa()</a></code>

        </dt>
        <dd>Calculates eigentaxa values based on SparCC - MCL generated clusters per sample. It is advised to run [cluster_taxa()] on the tidytacos object before running this function to add the clusters if you want to stray from any default parameters.</dd>
      </dl></div><div class="section level2">
      <h2 id="diversity-analysis">Diversity analysis<a class="anchor" aria-label="anchor" href="#diversity-analysis"></a></h2>

      <div class="section-desc"><p>Functions to add various diversity metrics to the tidytacos object.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="add_alpha.html">add_alpha()</a></code>

        </dt>
        <dd>Add alpha diversity measure</dd>
      </dl><dl><dt>

          <code><a href="add_alphas.html">add_alphas()</a></code>

        </dt>
        <dd>Add alpha diversity measures</dd>
      </dl><dl><dt>

          <code><a href="betas.html">betas()</a></code>

        </dt>
        <dd>Get beta diversity table</dd>
      </dl><dl><dt>

          <code><a href="add_ord.html">add_ord()</a></code>

        </dt>
        <dd>Add ordination</dd>
      </dl><dl><dt>

          <code><a href="add_copca.html">add_copca()</a></code>

        </dt>
        <dd>Add compositional principal components to the sample table</dd>
      </dl><dl><dt>

          <code><a href="add_tree.html">add_tree()</a></code>

        </dt>
        <dd>Construct a phylogeny from ASV sequences</dd>
      </dl><dl><dt>

          <code><a href="calculate_unifrac_distances.html">calculate_unifrac_distances()</a></code>

        </dt>
        <dd>Calculate unifrac distance matrix from a tidytacos object with a rooted tree</dd>
      </dl></div><div class="section level2">
      <h2 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a></h2>

      <div class="section-desc"><p>Functions to cluster the sample and taxa table.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="cluster_samples.html">cluster_samples()</a></code>

        </dt>
        <dd>Clusters samples into n clusters</dd>
      </dl><dl><dt>

          <code><a href="cluster_taxa.html">cluster_taxa()</a></code>

        </dt>
        <dd>Performs SparCC network analysis on a tidytacos object and then performs Markov Clustering on the network to annotate taxa of the largest clusters in the tidytacos object.</dd>
      </dl><dl><dt>

          <code><a href="add_sample_clustered.html">add_sample_clustered()</a></code>

        </dt>
        <dd>Add clustering-based sample order</dd>
      </dl></div><div class="section level2">
      <h2 id="network-analysis">Network analysis<a class="anchor" aria-label="anchor" href="#network-analysis"></a></h2>

      <div class="section-desc"><p>Functions to perform network inference, clustering and filtering the resulting clusters.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="network.html">network()</a></code>

        </dt>
        <dd>Perform network inference with SparCC on a tidytacos object, after dropping rare taxa. See [SpiecEasi::sparcc()].</dd>
      </dl><dl><dt>

          <code><a href="filter_network.html">filter_network()</a></code>

        </dt>
        <dd>Filters the output of [network()] to a minimal threshold and transforms to matrix for downstream clustering or heatplot visualization.</dd>
      </dl><dl><dt>

          <code><a href="cluster_network.html">cluster_network()</a></code>

        </dt>
        <dd>Performs Markov Clustering on a sparcc network matrix generated by [filter_network()]. and returns a tibble with clusters and taxon identifiers. Optionally, the network can be visualized.</dd>
      </dl></div><div class="section level2">
      <h2 id="datasets">Datasets<a class="anchor" aria-label="anchor" href="#datasets"></a></h2>

      <div class="section-desc"><p>Test datasets used as examples.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="urt.html">urt</a></code>

        </dt>
        <dd>Upper Respiratory Tract samples</dd>
      </dl><dl><dt>

          <code><a href="leaf.html">leaf</a></code>

        </dt>
        <dd>Phylosphere samples</dd>
      </dl></div><div class="section level2">
      <h2 id="miscellaneous">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous"></a></h2>

      <div class="section-desc"><p>Functions to tidy up the three tables or convert to other data structures.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="trim_asvs.html">trim_asvs()</a></code>

        </dt>
        <dd>Trim all sequences</dd>
      </dl><dl><dt>

          <code><a href="classify_taxa.html">classify_taxa()</a></code>

        </dt>
        <dd>(Re)classify amplicon sequences</dd>
      </dl><dl><dt>

          <code><a href="reset_ids.html">reset_ids()</a></code>

        </dt>
        <dd>Reset the taxon and sample IDs</dd>
      </dl><dl><dt>

          <code><a href="change_id_samples.html">change_id_samples()</a></code>

        </dt>
        <dd>Change sample IDs to a given expression</dd>
      </dl><dl><dt>

          <code><a href="change_id_taxa.html">change_id_taxa()</a></code>

        </dt>
        <dd>Change taxon IDs to a given expression</dd>
      </dl><dl><dt>

          <code><a href="remove_empty_samples.html">remove_empty_samples()</a></code>

        </dt>
        <dd>Removes empty samples from the tidytacos object</dd>
      </dl><dl><dt>

          <code><a href="tidy_count_to_matrix.html">tidy_count_to_matrix()</a></code>

        </dt>
        <dd>Convert counts tidy data frame to matrix</dd>
      </dl><dl><dt>

          <code><a href="rarefy.html">rarefy()</a></code>

        </dt>
        <dd>Rarefy the samples to a given number of reads</dd>
      </dl><dl><dt>

          <code><a href="test_taco.html">test_taco()</a></code>

        </dt>
        <dd>Create a tidytacos object for testing/example purporses</dd>
      </dl><dl><dt>

          <code><a href="taxonlist_per_condition.html">taxonlist_per_condition()</a></code>

        </dt>
        <dd>Return a list of taxon IDs per condition</dd>
      </dl><dl><dt>

          <code><a href="tacosum.html">tacosum()</a></code>

        </dt>
        <dd>Return some descriptive numbers</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Stijn Wittouck, Tim Van Rillaer, Wenke Smets, Sarah Lebeer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

