<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Taxa and sample quality control • tidytacos</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Taxa and sample quality control">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidytacos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/LebeerLab/tidytacos/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Taxa and sample quality control</h1>
                        <h4 data-toc-skip class="author">Wenke
Smets</h4>
            
            <h4 data-toc-skip class="date">2024-04-03</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/LebeerLab/tidytacos/blob/v1.0.3/vignettes/data-processing.Rmd" class="external-link"><code>vignettes/data-processing.Rmd</code></a></small>
      <div class="d-none name"><code>data-processing.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="taxa-and-sample-quality-control">Taxa and sample quality control<a class="anchor" aria-label="anchor" href="#taxa-and-sample-quality-control"></a>
</h2>
<p>2024-04-03</p>
<p>This tutorial walks you through some quality checks when first diving
into a dataset of taxonomic compositions. Note that this tutorial does
not include sequence quality filtering, as we are assuming this was
already done by for example DADA2. Here we will focus on filtering the
dataset to keep only bacterial reads, inspect blanks, and removing
samples with too few reads. How you handle these steps may depend on
your dataset, but here’s a demonstration to give you an idea of your
options.</p>
<div class="section level4">
<h4 id="setting-up">Setting up<a class="anchor" aria-label="anchor" href="#setting-up"></a>
</h4>
<p>We need only two packages: tidytacos (of course) and the tidyverse
set of packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://lebeerlab.github.io/tidytacos/">tidytacos</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="our-data">Our data<a class="anchor" aria-label="anchor" href="#our-data"></a>
</h4>
<p>The data for this tutorial is found in the tidytacos object ‘leaf’.
This data comes from a study where the phyllosphere microbiome of plants
is treated. After sampling, before DNA extraction a spike bacterium was
added to all samples to serve as an internal standard to determine
absolute abundances of the bacteria in our samples (as explained in
Smets et al. 2016).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tacosum.html">tacosum</a></span><span class="op">(</span><span class="va">leaf</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## n_samples    n_taxa   n_reads </span></span>
<span><span class="co">##        33      6054   4389396</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="filter-out-undesired-taxa">Filter out undesired taxa<a class="anchor" aria-label="anchor" href="#filter-out-undesired-taxa"></a>
</h4>
<p>In this study, we choose to sequence the V4 region of the 16S rRNA
gene to specifically study the bacterial fraction of our microbial
community, hence <code>filter_taxa</code> is used to discard any
non-bacterial reads such as mitochondria, chloroplasts, and archaea.
Using <code>tacosum</code> before and after the filtering helps us
understand that only 60% of our reads were bacterial. In phyllosphere
samples it is not uncommon to find a large fraction of chloroplasts
and/or mitochondria, plant cell organelles which contain a 16S rRNA
gene, hence the substantial loss of reads in this step.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">before</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tacosum.html">tacosum</a></span><span class="op">(</span><span class="va">leaf</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span></span>
<span></span>
<span><span class="va">leaf</span> <span class="op">&lt;-</span> <span class="va">leaf</span><span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/filter_taxa.html">filter_taxa</a></span><span class="op">(</span><span class="va">kingdom</span> <span class="op">==</span> <span class="st">"Bacteria"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/tacosum.html">tacosum</a></span><span class="op">(</span><span class="va">leaf</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## n_samples    n_taxa   n_reads </span></span>
<span><span class="co">##        33      5589   2634933</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">after</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/tacosum.html">tacosum</a></span><span class="op">(</span><span class="va">leaf</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="va">after</span><span class="op">/</span><span class="va">before</span></span></code></pre></div>
<pre><code><span><span class="co">##   n_reads </span></span>
<span><span class="co">## 0.6002951</span></span></code></pre>
<p>Depending on the database you used for assigning taxonomy, you may
need to adapt this code and use for example:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leaf</span> <span class="op">&lt;-</span> <span class="va">leaf</span><span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/filter_taxa.html">filter_taxa</a></span><span class="op">(</span><span class="va">class</span> <span class="op">!=</span> <span class="st">"Chloroplast"</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">class</span><span class="op">)</span><span class="op">)</span><span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/filter_taxa.html">filter_taxa</a></span><span class="op">(</span><span class="va">family</span> <span class="op">!=</span> <span class="st">"Mitochondria"</span><span class="op">|</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">family</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="read-numbers-and-blanks">Read numbers and blanks<a class="anchor" aria-label="anchor" href="#read-numbers-and-blanks"></a>
</h4>
<p>An important initial quality control is an investigation of the read
numbers per sample. The <code>add_total_count</code> function will add
total read counts per sample in the sample table. We can then visualize
the distribution of these read counts using ggplot.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leaf</span> <span class="op">&lt;-</span> <span class="va">leaf</span><span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/add_total_count.html">add_total_count</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">leaf</span><span class="op">$</span><span class="va">samples</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">total_count</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">sample</span>, color <span class="op">=</span> <span class="va">Plant</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="data-processing_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
<p>Unfortunately, one sample (bottom right corner) took up a lot of the
reads of this sequencing run, decreasing sampling depth of all other
samples. Nevertheless, most samples have more than 1000 reads, which we
consider worthwhile. Blanks are on the low end of the spectrum in
general, which is according to our expectation, but some blanks have
quite some reads which requires further investigation.</p>
<p>Let’s look at the composition of our blanks. We use the
<code>filter_samples</code> function to generate a tidytacos object with
only the blanks. The <code>tacoplot_stack</code> function returns a nice
stacked bar plot visualization of the most abundant taxa in the
blanks.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leaf1</span> <span class="op">&lt;-</span> <span class="va">leaf</span><span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/filter_samples.html">filter_samples</a></span><span class="op">(</span><span class="va">Plant</span> <span class="op">==</span> <span class="st">"Blank"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/tacoplot_stack.html">tacoplot_stack</a></span><span class="op">(</span><span class="va">leaf1</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="op">-</span><span class="fl">0.04</span>, label <span class="op">=</span> <span class="va">Plot</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2.5</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="op">-</span><span class="fl">0.06</span>, label <span class="op">=</span> <span class="va">total_count</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="data-processing_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
<p>Citrobacter is suspiciously prevalent and occurs in multiple blanks!
After checking its sequence, this ASV is the one that we spiked into our
samples and many of our blanks to determine absolute bacterial numbers
from the relative abundances. Hence, in this case, the high abundance of
Citrobacter in the blanks is a good thing. Besides Citrobacter none of
the contaminants are consistent nor do the read numbers of the blanks
without spike exceed 1000 per sample, which are all good signs for the
quality of our data.</p>
<p>We still need to set a minimum read number that a sample needs to
have to be included in our analyses. Samples with really low read counts
are too likely not being representative of the sampled microbiome by
poor amplification, contamination, and/or insufficient sequencing depth.
Usually, a minimum read number cut-off is quite an arbitrary decision,
hopefully somewhat informed by the nature of your data and results from
the blanks. In this case however, we have spiked our samples and several
of our blanks and we can determine a threshold that is actually informed
by the amount of contaminants in the blanks!</p>
</div>
<div class="section level4">
<h4 id="absolute-abundances">Absolute abundances<a class="anchor" aria-label="anchor" href="#absolute-abundances"></a>
</h4>
<p>We added equal amounts of a spike taxon (Citrobacter) to all of our
samples and some of the blanks. This helps us calculate the absolute
abundances of the 16S gene copies per sample using the function
<code>add_total_absolute_abundance</code>. First we need to determine
the spike taxon_id. Using <code>add_prevalence</code> will help us
determine which taxa are present in all samples. After comparing
taxonomy and/or sequence with the information we have of our spike, we
can identify the spike taxon_id, which we need for the
<code>add_total_absolute_abundance</code> function.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leaf</span> <span class="op">&lt;-</span> <span class="va">leaf</span><span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/add_prevalence.html">add_prevalence</a></span><span class="op">(</span>relative <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">leaf</span><span class="op">$</span><span class="va">taxa</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 10</span></span></span>
<span><span class="co">##   kingdom  phylum     class order family genus species taxon taxon_id prevalence</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> Bacteria Proteobac… Gamm… Ente… Enter… Citr… <span style="color: #BB0000;">NA</span>      TACG… t6           0.788 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> Bacteria Proteobac… Gamm… Ente… Enter… Buch… <span style="color: #BB0000;">NA</span>      TACG… t8           0.030<span style="text-decoration: underline;">3</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> Bacteria Proteobac… Beta… Burk… Oxalo… Mass… aurea/… TACG… t9           0.364 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> Bacteria Proteobac… Gamm… Ente… Enter… Buch… <span style="color: #BB0000;">NA</span>      TACG… t11          0.060<span style="text-decoration: underline;">6</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> Bacteria Proteobac… Alph… Sphi… Sphin… Sphi… <span style="color: #BB0000;">NA</span>      TACG… t12          0.515 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> Bacteria Actinobac… Acti… Fran… Geode… Blas… <span style="color: #BB0000;">NA</span>      TACG… t14          0.485</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leaf</span> <span class="op">&lt;-</span> <span class="va">leaf</span><span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/add_total_absolute_abundance.html">add_total_absolute_abundance</a></span><span class="op">(</span>spike_taxon <span class="op">=</span> <span class="st">"t6"</span>, spike_added <span class="op">=</span> <span class="va">added_spike_copies</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">leaf</span><span class="op">$</span><span class="va">samples</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Plant</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">total_absolute_abundance</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="data-processing_files/figure-html/unnamed-chunk-7-1.png" width="768"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">contaminants</span> <span class="op">&lt;-</span> <span class="va">leaf</span><span class="op">$</span><span class="va">samples</span><span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Plant</span> <span class="op">==</span> <span class="st">"Blank"</span><span class="op">)</span><span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">total_absolute_abundance</span><span class="op">)</span><span class="op">)</span><span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">total_absolute_abundance</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">leaf</span> <span class="op">&lt;-</span> <span class="va">leaf</span><span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/mutate_samples.html">mutate_samples</a></span><span class="op">(</span>potential_contamination <span class="op">=</span> <span class="va">contaminants</span><span class="op">$</span><span class="va">mean</span><span class="op">/</span><span class="va">total_absolute_abundance</span><span class="op">)</span><span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/filter_samples.html">filter_samples</a></span><span class="op">(</span><span class="va">potential_contamination</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/tacosum.html">tacosum</a></span><span class="op">(</span><span class="va">leaf</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## n_samples    n_taxa   n_reads </span></span>
<span><span class="co">##        16      5536   2551543</span></span></code></pre>
<p>We see that the estimated biomass in the samples is well above the
estimated biomass in the blanks, which is again a good sign! As we would
like to minimize our chances to have more than 10% of contamination in
the samples, we decided to remove samples that have less than 10x more
biomass than the mean biomass in the blanks. This results in a tidytacos
object without blanks and without 2 samples which were of such low
biomass that we considered them too sensitive to contamination. An
excellent starting point to start analyses.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Stijn Wittouck, Tim Van Rillaer, Wenke Smets, Sarah Lebeer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
