<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Tidytacos quick start guide • tidytacos</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Tidytacos quick start guide">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidytacos</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/LebeerLab/tidytacos/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="getting-started_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="getting-started_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><script src="getting-started_files/plotly-binding-4.10.4/plotly.js"></script><script src="getting-started_files/typedarray-0.1/typedarray.min.js"></script><link href="getting-started_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="getting-started_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><link href="getting-started_files/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="getting-started_files/plotly-main-2.11.1/plotly-latest.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Tidytacos quick start guide</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/LebeerLab/tidytacos/blob/master/vignettes/getting-started.Rmd" class="external-link"><code>vignettes/getting-started.Rmd</code></a></small>
      <div class="d-none name"><code>getting-started.Rmd</code></div>
    </div>

    
    
<p>This guide explains how to perform some common microbial sequencing
count data analysis tasks using tidytacos. We will illustrate those
using a dataset with human microbiome samples from the upper respiratory
tract (URT), taken from <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2017.02372/full" class="external-link">this
paper</a> by De Boeck et al. It contains nose as well as nasopharynx
samples. Most samples were taken using a swab method, but a minority was
taking with the aspirate method.</p>
<div class="section level2">
<h2 id="philosophy-of-the-package">Philosophy of the package<a class="anchor" aria-label="anchor" href="#philosophy-of-the-package"></a>
</h2>
<p>A tidytacos object is simply a list of three tables:</p>
<ul>
<li>
<strong>counts</strong>: These are the counts of reads for each
taxon (OTU/ASV/phylotype) in each sample. Each row represents such a
read count.</li>
<li>
<strong>samples</strong>: This table contains the sample metadata.
Each row represents a sample.</li>
<li>
<strong>taxa</strong>: This table contains the taxonomy and other
metadata for the taxa. Each row represents a taxon.</li>
</ul>
<p>The package is called tidytacos because each of the tables is tidy:
every row represents an observation and every column a variable (more on
data tidying can be found at <a href="https://r4ds.hadley.nz/data-tidy.html#sec-tidy-data" class="external-link uri">https://r4ds.hadley.nz/data-tidy.html#sec-tidy-data</a>).
For this quick start guide, we assume you are familiar with the
tidyverse (especially dplyr and ggplot).</p>
<p>The main differences with the popular phyloseq package are:</p>
<ul>
<li>The abundance table of tidytacos is in long format, while the
otu_table object of phyloseq is in wide (matrix) format. A tidy read
count table is more compact for very sparse data (such as microbiome
data), and also easier to handle and visualize within the tidyverse
framework.</li>
<li>All tables in tidytacos are data frames (technically tibbles). In
phyloseq, each table has its own special data type (i.e. the classes
“otu_table”, “tax_table” and “sample_data”). This makes it sometimes
difficult to access the data directly and do simple stuff with it. Also,
the otu_table and tax_table classes are based on the matrix type, while
the sample_data class is based on the dataframe type. This also makes
data manipulation sometimes unstraightforward.</li>
<li>In a phyloseq otu table, sometimes the rows are the taxa and
sometimes the columns. This can easily lead to errors.</li>
</ul>
</div>
<div class="section level2">
<h2 id="setting-up">Setting up<a class="anchor" aria-label="anchor" href="#setting-up"></a>
</h2>
<p>In case you haven’t installed tidytacos yet, it can be installed
using devtools:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"LebeerLab/tidytacos"</span><span class="op">)</span></span></code></pre></div>
<p>For this guide, we only need to load two packages: tidytacos (of
course) and the tidyverse set of packages.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Attaching core tidyverse packages</span> ──────────────────────── tidyverse 2.0.0 ──</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">dplyr    </span> 1.1.4     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">readr    </span> 2.1.5</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">forcats  </span> 1.0.0     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">stringr  </span> 1.5.1</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">ggplot2  </span> 3.5.2     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tibble   </span> 3.2.1</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">lubridate</span> 1.9.4     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tidyr    </span> 1.3.1</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">purrr    </span> 1.0.4     </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Conflicts</span> ────────────────────────────────────────── tidyverse_conflicts() ──</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">filter()</span> masks <span style="color: #0000BB;">stats</span>::filter()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">lag()</span>    masks <span style="color: #0000BB;">stats</span>::lag()</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use the conflicted package (<span style="color: #0000BB; font-style: italic;">&lt;http://conflicted.r-lib.org/&gt;</span>) to force all conflicts to become errors</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://lebeerlab.github.io/tidytacos/">tidytacos</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'tidytacos'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     everything</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following object is masked from 'package:tidyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     everything</span></span></code></pre></div>
<p>The first step will usually be the conversion our data into a usable
tidytacos format. If, for example, you used DADA2 as explained <a href="http://rstudio-pubs-static.s3.amazonaws.com/273862_8258982fdfac49a8ba9d6ed651b9b539.html" class="external-link">here</a>,
you can hand off the data to tidytacos using <code><a href="../reference/from_dada.html">from_dada()</a></code>.
You can also convert a phyloseq object to a tidytacos object using the
<code><a href="../reference/from_phyloseq.html">from_phyloseq()</a></code></p>
<p>function. More options to import and convert your data can be found
<a href="https://lebeerlab.github.io/tidytacos/reference/index.html#file-handling">here</a>.</p>
<p>A tidytacos object is read and stored as three sparse tables
(counts-, taxa- and samples.csv). The sample table needs atleast a
<code>sample_id</code> column, the taxa table needs a
<code>taxon_id</code> column with atleast one other rank name column and
the counts table needs a <code>sample_id</code>, <code>taxon_id</code>
and <code>counts</code> column. The taxa table of a default tidytacos
object has the domain, phylum, class, order, family, genus and species
columns. Species here has a special place as it is often ignored by some
functions unless specified, due to the low resolution of short read 16S
rRNA gene sequencing data.</p>
<p>To read in existing data from a folder called ‘my_data’ you would
run:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"tidytacos"</span>, <span class="st">"leaf"</span>, package <span class="op">=</span> <span class="st">"tidytacos"</span><span class="op">)</span></span>
<span><span class="co"># my_path &lt;- "local/path/my_data"</span></span>
<span><span class="va">taco</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_tidytacos.html">read_tidytacos</a></span><span class="op">(</span><span class="va">my_path</span><span class="op">)</span></span></code></pre></div>
<p>Where you replace my_path with an actual path to your data. The path
used here uses data in our package.</p>
<p>To write your data to a folder called “my_data_filtered” you can
run:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taco</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/write_tidytacos.html">write_tidytacos</a></span><span class="op">(</span><span class="st">"my_data_filtered"</span><span class="op">)</span></span></code></pre></div>
<p>However, our example dataset is available in the tidytacos package
and doesn’t need to be imported or converted. It is called “urt” and we
start by inspecting the samples table:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">urt</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 217</span></span>
<span><span class="co">#&gt; Columns: 9</span></span>
<span><span class="co">#&gt; $ run         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "20161207_ilke_urt100"<span style="color: #949494;">, </span>"20161207_ilke_urt100"<span style="color: #949494;">, </span>"20161207_…</span></span>
<span><span class="co">#&gt; $ condition   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "CON"<span style="color: #949494;">, </span>"CON"<span style="color: #949494;">, </span>"CON"<span style="color: #949494;">, </span>"CON"<span style="color: #949494;">, </span>"CON"<span style="color: #949494;">, </span>"CON"<span style="color: #949494;">, </span>"CON"<span style="color: #949494;">, </span>"CON"<span style="color: #949494;">, </span>"C…</span></span>
<span><span class="co">#&gt; $ participant <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "CON100"<span style="color: #949494;">, </span>"CON100"<span style="color: #949494;">, </span>"CON10"<span style="color: #949494;">, </span>"CON10"<span style="color: #949494;">, </span>"CON11"<span style="color: #949494;">, </span>"CON11"<span style="color: #949494;">, </span>"C…</span></span>
<span><span class="co">#&gt; $ location    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "NF"<span style="color: #949494;">, </span>"N"<span style="color: #949494;">, </span>"NF"<span style="color: #949494;">, </span>"NF"<span style="color: #949494;">, </span>"NF"<span style="color: #949494;">, </span>"N"<span style="color: #949494;">, </span>"NF"<span style="color: #949494;">, </span>"NF"<span style="color: #949494;">, </span>"N"<span style="color: #949494;">, </span>"NF"<span style="color: #949494;">, </span>"…</span></span>
<span><span class="co">#&gt; $ method      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "S"<span style="color: #949494;">, </span>"S"<span style="color: #949494;">, </span>"A"<span style="color: #949494;">, </span>"S"<span style="color: #949494;">, </span>"S"<span style="color: #949494;">, </span>"S"<span style="color: #949494;">, </span>"S"<span style="color: #949494;">, </span>"S"<span style="color: #949494;">, </span>"S"<span style="color: #949494;">, </span>"S"<span style="color: #949494;">, </span>"S"<span style="color: #949494;">, </span>"S"…</span></span>
<span><span class="co">#&gt; $ plate       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 3<span style="color: #949494;">, </span>3<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1<span style="color: #949494;">, </span>1…</span></span>
<span><span class="co">#&gt; $ passes_qc   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> TRUE<span style="color: #949494;">, </span>TRUE<span style="color: #949494;">, </span>TRUE<span style="color: #949494;">, </span>TRUE<span style="color: #949494;">, </span>TRUE<span style="color: #949494;">, </span>TRUE<span style="color: #949494;">, </span>TRUE<span style="color: #949494;">, </span>TRUE<span style="color: #949494;">, </span>TRUE<span style="color: #949494;">, </span>TRUE…</span></span>
<span><span class="co">#&gt; $ sample      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "CON100-NF-S"<span style="color: #949494;">, </span>"CON100-N-S"<span style="color: #949494;">, </span>"CON10-NF-A"<span style="color: #949494;">, </span>"CON10-NF-S"<span style="color: #949494;">, </span>"…</span></span>
<span><span class="co">#&gt; $ sample_id   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "s1"<span style="color: #949494;">, </span>"s2"<span style="color: #949494;">, </span>"s3"<span style="color: #949494;">, </span>"s4"<span style="color: #949494;">, </span>"s5"<span style="color: #949494;">, </span>"s6"<span style="color: #949494;">, </span>"s7"<span style="color: #949494;">, </span>"s8"<span style="color: #949494;">, </span>"s9"<span style="color: #949494;">, </span>"s10…</span></span></code></pre></div>
<p>We then have a quick look at the total number of samples, ASVs, and
reads in the tidytacos object:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tacosum.html">tacosum</a></span><span class="op">(</span><span class="va">urt</span><span class="op">)</span></span>
<span><span class="co">#&gt; n_samples    n_taxa   n_reads </span></span>
<span><span class="co">#&gt;       217      1957   3873478</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="making-a-stacked-barplot-of-a-subset-of-samples">Making a stacked barplot of a subset of samples<a class="anchor" aria-label="anchor" href="#making-a-stacked-barplot-of-a-subset-of-samples"></a>
</h2>
<p>We can very easily create a plot to explore a subset of our samples
(e.g., only nose samples taken with the swab method) in the following
way:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">urt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/filter_samples.html">filter_samples</a></span><span class="op">(</span><span class="va">location</span> <span class="op">==</span> <span class="st">"N"</span>, <span class="va">method</span> <span class="op">==</span> <span class="st">"S"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/tacoplot_stack.html">tacoplot_stack</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started_files/figure-html/stack-plot-1.png" width="768"></p>
<p>The <code>filter_samples</code> function does what it says: filtering
samples. It will also delete taxa from the taxa table that have zero
total reads in the remaining samples. The <code>tacoplot_stack</code>
function returns a nice stacked bar plot visualization of the most
abundant taxa in our samples.</p>
</div>
<div class="section level2">
<h2 id="subsetting-dataset">Subsetting dataset<a class="anchor" aria-label="anchor" href="#subsetting-dataset"></a>
</h2>
<p>Our next question for this dataset is to what extent nose and
nasopharynx are linked from a microbiological point of view. To get an
idea we can first visualize this:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">urt_s</span> <span class="op">&lt;-</span> <span class="va">urt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/filter_samples.html">filter_samples</a></span><span class="op">(</span><span class="va">method</span> <span class="op">==</span> <span class="st">"S"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tacoplot_stack.html">tacoplot_stack</a></span><span class="op">(</span><span class="va">urt_s</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="op">-</span><span class="fl">0.02</span>,color<span class="op">=</span><span class="va">location</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started_files/figure-html/subset-plot-1.png" width="768"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tacoplot_stack.html">tacoplot_stack</a></span><span class="op">(</span><span class="va">urt_s</span>, x <span class="op">=</span> <span class="va">location</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">participant</span>, nrow <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in tacoplot_stack(urt_s, x = location): Sample labels not unique,</span></span>
<span><span class="co">#&gt; samples are aggregated.</span></span></code></pre></div>
<p><img src="getting-started_files/figure-html/facet-plot-1.png" width="768"></p>
<p>First, we make sure to include only swab samples. We can then
visualize the data in stacked bar plots. The tacoplot functions consist
of ggplot function calls and therefore geoms and other ggplot features
can be added.</p>
</div>
<div class="section level2">
<h2 id="alpha-diversity">Alpha diversity<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a>
</h2>
<p>To explore alpha diversity, let’s create a rarefied version of the
dataset:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">urt_rar</span> <span class="op">&lt;-</span> <span class="va">urt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/add_total_count.html">add_total_count</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/filter_samples.html">filter_samples</a></span><span class="op">(</span><span class="va">total_count</span> <span class="op">&gt;=</span> <span class="fl">2000</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="fl">2000</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/add_alphas.html">add_alphas</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>We can visualize alpha diversities of nose versus the
nasopharynx:</p>
<p>The <code>add_total_count</code> function will add total sample read
numbers to the sample table.</p>
<p>The <code>rarefy</code> function will randomly subsample all samples
n times. It only works if the read count of each sample equals or
exceeds n. For determining ASV richness, we chose to rarefy first, but
this may depend on your data.</p>
<p>The <code>add_alpha</code> function can be used to add several
metrics of alpha diversity to the sample table or in this cas we use the
add_alphas to add all the currently supported metrics.</p>
<p>Next we will plot the resulting alpha diversities and group per
location of sampling:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">urt_rar</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/tacoplot_alphas.html">tacoplot_alphas</a></span><span class="op">(</span><span class="va">location</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 4 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_ydensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 4 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_point()`).</span></span></code></pre></div>
<p><img src="getting-started_files/figure-html/plot-alpha-1.png" width="672"></p>
<p>The tacoplot_alphas function will use any alpha metric that has been
added by add_alpha(s) which it can find in the sample table.</p>
</div>
<div class="section level2">
<h2 id="principal-coordinates-analysis">Principal coordinates analysis<a class="anchor" aria-label="anchor" href="#principal-coordinates-analysis"></a>
</h2>
<p>We would like to address the differences between nose and
nasopharynx. We’re also more interested in genera than ASVs. A PCA might
offer insight:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">urt_genus</span> <span class="op">&lt;-</span> <span class="va">urt</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/filter_samples.html">filter_samples</a></span><span class="op">(</span><span class="va">method</span> <span class="op">==</span> <span class="st">"S"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/aggregate_taxa.html">aggregate_taxa</a></span><span class="op">(</span>rank <span class="op">=</span> <span class="st">"genus"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/tacoplot_ord_ly.html">tacoplot_ord_ly</a></span><span class="op">(</span><span class="va">urt_genus</span>,<span class="va">location</span>, samplenames <span class="op">=</span> <span class="va">sample</span>, dim <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in remove_empty_samples(.): Removed 2 empty samples.</span></span></code></pre></div>
<div class="plotly html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"visdat":{"87c369892d0":["function () ","plotlyVisDat"]},"cur_data":"87c369892d0","attrs":{"87c369892d0":{"x":{},"y":{},"z":{},"text":{},"hovertemplate":"<i>%{text}<\/i>","color":{},"colors":["#e8e8e8","#a6cee3","#1f78b4","#b2df8a","#33a02c","#fb9a99","#e31a1c","#fdbf6f","#ff7f00","#cab2d6","#6a3d9a","#ffff99","#b15928"],"alpha_stroke":1,"sizes":[10,100],"spans":[1,20],"type":"scatter3d","mode":"markers","inherit":true}},"layout":{"margin":{"b":40,"l":60,"t":25,"r":10},"title":"PCOA plot","yaxis":{"zeroline":false},"xaxis":{"zeroline":false},"scene":{"xaxis":{"title":"ord1"},"yaxis":{"title":"ord2"},"zaxis":{"title":"ord3"}},"hovermode":"closest","showlegend":true},"source":"A","config":{"modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"data":[{"x":[0.30341154338864801,0.30871026021531645,0.32888540179799808,0.26931240929981959,0.14516623447026908,0.27269486187706843,0.28549254460240464,0.15461057120456065,0.28211922635021719,0.20968184473820323,0.26693667586661074,0.18109671348650119,0.13050541590559397,-0.28064991271259515,0.30987774562552811,0.23849876410573276,0.0050494262673653686,0.10983329076897586,-0.12362105236862574,0.30518938596841849,0.26190273632213007,-0.19975454986383154,0.33157878179664491,0.30558481470630305,0.24607066451278614,0.31098700124221251,0.28431025005408944,0.22398108608292949,0.16730570305647927,-0.55444308500984041,-0.44161029645235683,0.19904687305731,0.085473720608332457,-0.076035038217665316,0.29012903366366372,-0.10352133214130506,0.23233625155367971,0.23796726059789022,0.32187177542591666,0.31843188461941246,0.26786528050156616,-0.13846711064572237,0.22661371916470757,0.23248686140693442,-0.24679001830455555,-0.042308519213876197,0.26155235621277245,0.32253061616677176,0.21744769925393331,0.28315463376440619,-0.011896882740013841,-0.0071816381812001917,0.2106786578359926,0.05345292305933852,0.11749092868934476,0.12329283411672837,-0.13073182856582172,0.23751814333371929,0.089388800226861506,0.26197249415934115,0.2754085014244978,0.29721754417328955,0.10165493982617903,0.01239099356624994,-0.46475886465767557,-0.090860238783206052,0.12482062256394041,0.33021240536202867,0.3115676870889712,0.27213975151762049,0.19986654883420293,0.18024312220303804,0.24119180979966706,0.29921308534206886,-0.099609324638421995,0.30232213779728628,0.17968948306642485,0.28947641423461784,-0.43186034851029231,0.25606541325743898,0.083656336613379567,0.31850652128902252,0.34466523511156338,0.3005848668741643,0.062306016425882946],"y":[0.015994020150006488,-0.12770130536009902,-0.13362202366167866,-0.12341325713736928,0.16245853450151182,0.03226568243169825,0.048511327972572696,0.12871990143147929,0.077547923427293317,0.10025247001949389,0.016656416346642482,0.1120049739785436,0.18719558192853614,-0.41946544596181745,-0.014617207781622536,-0.19616560971476113,0.21638399518818241,-0.066322513493350818,0.11655413518232181,-0.23042959200435262,-0.10362771315166665,-0.42622791420729667,-0.23769185283721367,0.021912431828529619,-0.066085715201869497,-0.068958931982750729,-0.072369025820406732,0.061554077739393039,0.089821637187298936,-0.2399195177624158,-0.30469180932307116,0.074648500035206178,0.090039115389927579,0.047472620588891692,-0.24818077677309677,0.2011157117964506,-0.16637305011165213,-0.062843911484856796,-0.22019856887652267,-0.19571945991289483,-0.098861170159512374,0.36105290102898829,0.0040308844102098932,-0.24696630297662733,-0.41172851279455519,0.066602642972064238,-0.11101098735471038,-0.13232419146708174,-0.083217157118953916,-0.25366353950476589,0.30587197727188659,-0.33979582735591313,0.014666390314942922,0.22749928529005722,0.14783443817600952,-0.077250035040929435,0.34493904382139862,-0.19865000244474029,0.10964318579233001,-0.092785694463812213,-0.085880918791293795,-0.20039697646067742,0.24489343398054475,0.30934024610530303,-0.44763962524042072,0.1185541922952823,0.016936854355477059,-0.14225445176050563,-0.1336141678496871,0.032194133798338923,0.10976890558044039,0.15514271872143492,0.042666534851719347,-0.26712879137338996,0.30792181381200051,0.0052628461784477304,0.16643651412778279,0.036196389631989699,-0.39253658728031343,-0.092719326724860454,0.19938493865069454,-0.060429677929527068,-0.066703892006250154,-0.15811707267821515,0.25253982893322319],"z":[0.030657219546176718,-0.12154304925168853,-0.11723240494871985,-0.09307778372780387,0.23497958695891358,0.097848496092156137,0.076270079597061516,0.092330152370852736,0.13250714824806695,0.14538325671324942,0.070871067649394145,0.098839595091712598,0.13055500412522153,0.15583237703202138,0.055184588326401364,-0.19375630435547098,-0.013595422385894639,0.055895977261741878,0.054626496778292899,-0.22136453309522305,-0.060595101337742117,0.12551542550145278,-0.23529376768232299,0.064521427589244285,-0.14540945143103845,-0.074016662278670597,-0.018535945623379339,0.10128469655649706,0.14748013692976736,0.1099572231930306,0.2312576691761366,0.0827878253318583,0.16796992303082187,0.052423183766951598,-0.25049688149355159,-0.051563002129028035,-0.21048970743367881,-0.027504255366472101,-0.20532340638255034,-0.16208363801564457,-0.016960445859432884,0.2328741244530925,0.03324747076228611,-0.11540972416507947,0.18653709657380138,0.17761238257858411,-0.14419548788393088,-0.10702688365713926,-0.063681701642644889,-0.23953052345015297,0.23914153657597409,-0.024084497947246413,0.042920118363055211,0.22134628814631332,0.083535189359979597,-0.10313144336370106,-0.030683898588674624,-0.11536255624227501,0.19900562116563986,-0.060964954617444451,-0.049043942151082834,-0.20711960657466269,0.21605687338252555,0.27965953615343486,0.17068324533987603,-0.18484085329845687,-0.099255069208518343,-0.10313337191596217,-0.069258824503058244,0.076291607310057449,0.089768459034422488,0.21482068082962985,0.1133751222841715,-0.23261824637494824,0.18283471407952923,0.078359788003014297,0.20039554334085008,0.089975357105574791,0.21783566372162017,-0.13291688511399216,0.070425135593728511,-0.034736670343981708,-0.008421184143274963,-0.13111533234526407,0.28606161968698635],"text":["CON100-N-S","CON11-N-S","CON13-N-S","CON14-N-S","CON15-N-S","CON16-N-S","CON17-N-S","CON18-N-S","CON19-N-S","CON20-N-S","CON21-N-S","CON22-N-S","CON23-N-S","CON24-N-S","CON25-N-S","CON26-N-S","CON27-N-S","CON28-N-S","CON29-N-S","CON30-N-S","CON31-N-S","CON33-N-S","CON35-N-S","CON36-N-S","CON37-N-S","CON38-N-S","CON39-N-S","CON40-N-S","CON41-N-S","CON42-N-S","CON43-N-S","CON44-N-S","CON45-N-S","CON46-N-S","CON47-N-S","CON48-N-S","CON49-N-S","CON50-N-S","CON51-N-S","CON52-N-S","CON53-N-S","CON54-N-S","CON55-N-S","CON56-N-S","CON57-N-S","CON59-N-S","CON60-N-S","CON61-N-S","CON62-N-S","CON63-N-S","CON64-N-S","CON65-N-S","CON66-N-S","CON67-N-S","CON68-N-S","CON69-N-S","CON70-N-S","CON71-N-S","CON72-N-S","CON74-N-S","CON75-N-S","CON76-N-S","CON77-N-S","CON78-N-S","CON79-N-S","CON80-N-S","CON81-N-S","CON82-N-S","CON83-N-S","CON84-N-S","CON85-N-S","CON86-N-S","CON87-N-S","CON88-N-S","CON89-N-S","CON90-N-S","CON91-N-S","CON92-N-S","CON93-N-S","CON94-N-S","CON95-N-S","CON96-N-S","CON97-N-S","CON98-N-S","CON99-N-S"],"hovertemplate":["<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>"],"type":"scatter3d","mode":"markers","name":"N","marker":{"color":"rgba(232,232,232,1)","line":{"color":"rgba(232,232,232,1)"}},"textfont":{"color":"rgba(232,232,232,1)"},"error_y":{"color":"rgba(232,232,232,1)"},"error_x":{"color":"rgba(232,232,232,1)"},"line":{"color":"rgba(232,232,232,1)"},"frame":null},{"x":[0.34336977636054317,-0.5614111780114337,0.30469423217741204,0.042284768458898558,0.038723197540078094,-0.43510119561738486,0.0051701009847972199,-0.04312444847885042,-0.00035223624235005915,-0.16354569153146253,0.014688865115493067,-0.1848211366424049,-0.25397832309544527,-0.15576961433030947,-0.039879005815872261,-0.23923320455905192,-0.11350835752905918,0.19163669266785011,-0.35708701177138169,-0.15159401181936819,-0.58339267518219207,0.13811937957143963,-0.14355007640675099,-0.3830073657496435,-0.27563264852460195,-0.15552669438780531,-0.31593801202418476,-0.40545255065096342,-0.38842205027485743,-0.52197121135183289,-0.40399213435144571,-0.20798284442660628,-0.54249524705864571,-0.0066141373023957525,-0.43716111053376883,-0.44014447481693419,0.058792247360318892,0.075795207103125914,-0.13873506237743319,0.27770520586244801,-0.083045559717719042,-0.10753169684632576,-0.40176930311499737,-0.077655368112604664,0.14658019059705166,-0.39163572156130799,0.02682593958520162,-0.24953724472262634,-0.31333235695737799,-0.11326096127708594,0.19989127675581489,-0.36494679252345741,-0.30805678826368382,-0.20138587576258324,-0.14579247983571136,0.018968304249314488,0.27953579521416339,-0.16942913436700979,-0.39556290480178069,0.028320147649040003,-0.48926637297272063,0.29879058325589025,0.32613039115143244,-0.39818201768277084,0.29539335685719659,0.053087877182980997,-0.44110319529738534,-0.40454260740342401,-0.44562944110628311,0.29809170367739496,-0.17534132782835907,-0.41432734658323672,-0.53739973499792737,-0.13164653836491222,-0.05056102372715747,-0.22910159632742336,0.27751090616671609,0.25984096117431998,-0.35463929579294168,-0.028223075423419768,0.090635653004231681,-0.53915245873844042,-0.1548600972503944,0.23266340261542487,-0.15312952311946917,0.093666398902270492,0.057792695461879387,-0.12476140785261634,-0.050211913425011807,0.093931671027778371,0.19227746359449102,0.33964733742414954,-0.13860757617365668,-0.093546963357922869,-0.54255781309092488],"y":[-0.1124513347728866,-0.19418162272987827,-0.086736389104480005,0.20396308235226024,-0.4096755216608835,-0.24252382778435994,0.24820509479619549,0.24016351120566104,0.34528649156763569,-0.024009020247439197,0.29360791787157497,0.059840191796498597,0.021228294806240822,0.23093433647071032,0.28665554498904505,0.20904630390572287,-0.41330596660976343,0.10908337092985777,0.22730133905445823,0.23559357639236697,-0.18172319067943554,0.21391307751300545,0.39920827344262988,0.27405101279597205,0.036549526536972382,-0.33517697957336995,0.12032375072790986,0.31282805301873601,0.26053794636507105,-0.42736484096557725,0.25766297703529417,0.22670857302485412,-0.46744141947496609,0.17372840430741929,-0.0016206855370020389,-0.22468187778554552,0.11748642689613757,0.1259657513113861,-0.0072404075560751693,-0.24750064667564978,0.31114523102184144,0.01892847001773974,-0.1252528532120962,0.15593177291997729,-0.00059195868607358907,0.079096311778401859,0.13389398114160081,0.01731575974949336,0.036677415126707039,-0.41814053498846876,-0.11418204975055901,-0.089374558231921988,0.25584418607142578,0.032376424987018607,0.17619635360249394,0.064847846614432728,-0.24985100512508543,0.29063938569073039,0.24097060022475658,-0.0045132502142588658,-0.42497529078294116,-0.18376698302102537,-0.29297727298432819,0.27036568532824484,-0.17963934940369791,0.1597686627544351,-0.43508248968180541,0.30333487940567244,-0.40095853979099683,-0.22397185441745118,0.028432108717778706,0.27301305063691833,-0.25192228274964545,-0.14863114683187925,0.28767821805764376,0.074864877151857939,0.0060958860489875172,-0.1886691465601833,0.27888880861397214,0.22101482403341646,-0.042421325659174634,-0.46413621582258513,0.29015153113790548,0.10727578426205071,0.24275553953463777,0.22796838264337294,0.18340649032189624,-0.16582617637048122,0.0366294737987592,0.097474294487175428,0.096272282039788326,-0.097037775494073358,0.20593515121014014,0.31202316831136773,-0.431718921507253],"z":[-0.083103138298609769,0.095712819709286551,-0.082649541324748804,0.25385269435306707,-0.042002200778759213,-0.0071475008516020833,0.28273282186253795,0.17171528504872699,0.21851431059680632,-0.10394042408582387,0.27151350656395973,-0.081351201306142928,-0.16588590739359413,-0.10992387618505012,0.074970979297840676,-0.022474175202729901,0.098561757201241024,0.082985109455968098,-0.45147087980696687,-0.046604996091845444,0.011971184371993042,0.29007078423576199,0.18620442481161956,-0.42568599312277089,-0.18710507009240177,0.20493154097654664,-0.24624356317503932,-0.38869169295673422,-0.44689046791020237,0.17134736047213775,-0.46275378583055921,-0.25455776714135592,0.17787053899856511,0.029606315767253269,0.01038957404872012,0.14181503380125046,0.17102965644366139,0.22905493148867542,-0.27049044286143009,-0.2450164815902745,0.11155410825780511,0.034470783434045943,-0.069222381340145495,-0.025599772437966566,0.11098597083138589,0.14959973269317697,-0.11814088411050883,-0.15404574409633309,-0.18668482499891276,0.14484247659891331,-0.00049985041647800804,-0.0031348732968438627,-0.13484646114840168,-0.10463431846480263,-0.097164475413215545,0.044082830932406711,-0.24154083037809437,0.14424850726336599,-0.4449040134082306,0.16582033932254742,0.17283228605379727,-0.11223759538088837,-0.29800661604508827,-0.43950341813396687,-0.12926297685125115,0.17991134551231841,0.18164629550861494,-0.40281955902325922,0.18183016091189674,-0.23715820112797212,-0.058598400158587888,-0.46292437870545783,0.040291172777713664,0.29910340855959028,0.30910616576884609,-0.15931981028900177,0.14001472843846954,-0.045666131979172783,-0.38174075727087636,0.067366329013706205,-0.10644772494076467,0.17624313331058034,0.03688161698416606,0.15199248339321914,-0.016361906778599913,0.27309108556065054,0.15292200398055938,0.23743612283412061,0.01453421348785245,0.03988055398612541,0.083722481347741948,-0.056702973605303471,-0.19205118051830572,0.06554151600993928,0.1760383736645337],"text":["CON100-NF-S","CON10-NF-S","CON11-NF-S","CON12-NF-S","CON13-NF-S","CON14-NF-S","CON15-NF-S","CON16-NF-S","CON17-NF-S","CON18-NF-S","CON19-NF-S","CON1-NF-S","CON20-NF-S","CON21-NF-S","CON22-NF-S","CON23-NF-S","CON24-NF-S","CON25-NF-S","CON26-NF-S","CON27-NF-S","CON29-NF-S","CON2-NF-S","CON30-NF-S","CON31-NF-S","CON32-NF-S","CON33-NF-S","CON35-NF-S","CON36-NF-S","CON37-NF-S","CON38-NF-S","CON39-NF-S","CON3-NF-S","CON40-NF-S","CON41-NF-S","CON42-NF-S","CON43-NF-S","CON44-NF-S","CON45-NF-S","CON46-NF-S","CON47-NF-S","CON48-NF-S","CON49-NF-S","CON4-NF-S","CON50-NF-S","CON51-NF-S","CON52-NF-S","CON53-NF-S","CON54-NF-S","CON55-NF-S","CON57-NF-S","CON58-NF-S","CON59-NF-S","CON5-NF-S","CON60-NF-S","CON61-NF-S","CON62-NF-S","CON63-NF-S","CON64-NF-S","CON65-NF-S","CON66-NF-S","CON67-NF-S","CON68-NF-S","CON69-NF-S","CON70-NF-S","CON71-NF-S","CON72-NF-S","CON73-NF-S","CON74-NF-S","CON75-NF-S","CON76-NF-S","CON77-NF-S","CON78-NF-S","CON79-NF-S","CON7-NF-S","CON80-NF-S","CON81-NF-S","CON83-NF-S","CON84-NF-S","CON85-NF-S","CON86-NF-S","CON87-NF-S","CON88-NF-S","CON89-NF-S","CON8-NF-S","CON90-NF-S","CON91-NF-S","CON92-NF-S","CON93-NF-S","CON94-NF-S","CON95-NF-S","CON96-NF-S","CON97-NF-S","CON98-NF-S","CON99-NF-S","CON9-NF-S"],"hovertemplate":["<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>","<i>%{text}<\/i>"],"type":"scatter3d","mode":"markers","name":"NF","marker":{"color":"rgba(177,89,40,1)","line":{"color":"rgba(177,89,40,1)"}},"textfont":{"color":"rgba(177,89,40,1)"},"error_y":{"color":"rgba(177,89,40,1)"},"error_x":{"color":"rgba(177,89,40,1)"},"line":{"color":"rgba(177,89,40,1)"},"frame":null}],"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.20000000000000001,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script><p>The <code>aggregate_taxa</code> function merges all rows of the taxa
table together at a specified taxonomic level, in this case the genus
level. As for all tidytacos functions, all other tables in the tidytacos
object are adjusted accordingly.</p>
<p>The <code>tacoplot_ord_ly</code> function will determine relative
abundances of taxa in the samples, and will then use Bray-Curtis
dissimilarities to ordinate samples in a 2-dimensional space based on
their taxonomic composition. The plotly “_ly” addition makes the plot
interactive, which is really nice for exploratory work. This also works
for other plot functions.</p>
</div>
<div class="section level2">
<h2 id="relation-between-community-composition-and-variables">Relation between community composition and variables<a class="anchor" aria-label="anchor" href="#relation-between-community-composition-and-variables"></a>
</h2>
<p>A logical next question is to what extent the niche (nose versus
nasopharynx) determines the community composition variability. Let’s not
forget that everyone has their unique microbiome and include the
variable “participant” in the model.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/perform_adonis.html">perform_adonis</a></span><span class="op">(</span><span class="va">urt_genus</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"participant"</span>, <span class="st">"location"</span><span class="op">)</span>, by<span class="op">=</span><span class="st">"margin"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Permutation test for adonis under reduced model</span></span>
<span><span class="co">#&gt; Marginal effects of terms</span></span>
<span><span class="co">#&gt; Permutation: free</span></span>
<span><span class="co">#&gt; Number of permutations: 999</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; adonis2(formula = as.formula(paste("counts_matrix", formula_RHS, sep = " ~ ")), data = metadata, permutations = permutations, by = "margin")</span></span>
<span><span class="co">#&gt;              Df SumOfSqs      R2       F Pr(&gt;F)    </span></span>
<span><span class="co">#&gt; participant  97   31.076 0.64174  1.8848  0.001 ***</span></span>
<span><span class="co">#&gt; location      1    3.066 0.06331 18.0372  0.001 ***</span></span>
<span><span class="co">#&gt; Residual     81   13.768 0.28432                   </span></span>
<span><span class="co">#&gt; Total       179   48.424 1.00000                   </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre></div>
<p>The <code>perform_adonis</code> function will perform a PERMutational
ANOVA to determine the effect of sample variables on the Bray-Curtis
dissimilarities of the communities. The result shows that participant is
a really important contributor to community composition (R squared =
0.65). Furthermore, there are consistent significant differences between
the communities of nose and nasopharynx (R squared = 0.06).</p>
</div>
<div class="section level2">
<h2 id="differential-abundance-analysis">Differential abundance analysis<a class="anchor" aria-label="anchor" href="#differential-abundance-analysis"></a>
</h2>
<p>Next, we would like to know which of the 20 most abundant genera, are
significantly more abundant in the nasopharynx compared to the nose.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">urt_genus</span> <span class="op">&lt;-</span> <span class="va">urt_genus</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/add_codifab.html">add_codifab</a></span><span class="op">(</span><span class="va">location</span>, max_taxa <span class="op">=</span> <span class="fl">40</span><span class="op">)</span></span>
<span><span class="va">urt_genus</span><span class="op">$</span><span class="va">taxon_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">urt_genus</span><span class="op">$</span><span class="va">taxon_pairs</span>, <span class="va">wilcox_p</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tacoplot_codifab.html">tacoplot_codifab</a></span><span class="op">(</span><span class="va">urt_genus</span>, <span class="va">NF_vs_N</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting-started_files/figure-html/codifab-1.png" width="768"></p>
<p>The <code>add_codifab</code> function will add a table called
taxon_pairs to the tidytacos object, with for each pair of a taxon and a
reference taxon, the differential abundance of the taxon between the two
conditions (with respect to the reference taxon).</p>
<p>The <code>codifab_plot</code> function returns a plot to visualize
differential abundance of taxa between conditions, compared to all other
taxa as references. We can observe that <em>Haemophilus</em> is most
likely to be typical for the Nasopharynx, whereas <em>Anaerococcus</em>
is most likely to be typical for the nose.</p>
<p>Of note, there are many differential abundance analysis methods out
there and none of them are perfect. Interpret your results with
care.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Stijn Wittouck, Tim Van Rillaer, Wenke Smets, Sarah Lebeer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
